substrate/pkg/controller/substrate/cluster/addons/utils.go

```diff
@@ -0,0 +1,78 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addons
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"syscall"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
+	"github.com/awslabs/kit/substrate/pkg/controller/substrate/cluster"
+	"github.com/awslabs/kit/substrate/pkg/utils/discovery"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+)
+
+var (
+	ErrWaitingForSubstrateEndpoint = fmt.Errorf("waiting for substrate cluster endpoint to be available")
+)
+
+const (
+	kubeAdminFilePath = "etc/kubernetes/admin.conf"
+)
+
+func KubeClientFor(ctx context.Context, substrate *v1alpha1.Substrate) (*kubernetes.Clientset, error) {
+	if substrate.Status.Cluster.Address == nil {
+		return nil, ErrWaitingForSubstrateEndpoint
+	}
+	// check if the kube-admin file for the cluster exists
```

Comment:
Will this race? Is there a way that you can inject a client as a dependency that can be shared between callers?

Reply:
> Will this race?

You mean race between controllers? Right now each controller for addon gets their own kubeclient.

>Is there a way that you can inject a client as a dependency to your other controllers that can be shared between callers?

I looked into this to be able to pass the same client to all controllers, but currently the way its designed is all the controllers are started in `substrate/controller.go` makes it tough to achieve. Other alternative can be have a wrapper which wraps all the addon controllers, and calls create only when kubeclient is created and can be shared between controllers